Boost: Add toggle UI with notice for Cache and trigger setup/cleanup

dilirity opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

Part of https://github.com/Automattic/jetpack/issues/35054

I have intentionally not styled the retry button and have not added content for (some) errors, as we haven't finalized those. During our weekly meeting, we discussed the option to have the "run setup" button inside the notice. The current implementation, is just for debugging. Will be addressed in a follow up PR.

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
* UI: Add a toggle to enable/disable the feature;
* Backend (PHP): Runs setup when feature is enabled;
* Backend (PHP): Runs cleanup when feature is disabled (cleans up advanced-cache.php and removes WP_CACHE from wp-config.php);
* UI: Add a notice that shows the last setup error. Maybe it would be a good idea to list all errors instead of just one?
* UI: Add a button to retry setup. Temporary for now, as we also need to add content for the various errors we might get during setup.
* React Data Sync Client: Add a way to invalidate queries. Used to invalidate the error data store, when the feature is enabled (to avoid showing the same error again);
* Backend (PHP): Add activate/deactivate traits, that are used when the feature is toggled on or off. This helps us run things on activation/deactivation (like setup and cleanup mentioned above). We should revisit existing code that runs code on activation/deactivation this in another way and clean it up.

<details><summary>Preview</summary>

![CleanShot 2024-02-13 at 14 06 05](https://github.com/Automattic/jetpack/assets/11799079/4743e030-2d7d-49bb-9187-572646b1cf9b)

</details>

 Since the feature is under a feature flag, don't forget to include `define( 'BOOST_CACHE', true );` in your `wp-config.php`.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

p1HpG7-r0o-p2

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

no

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Setup Boost;
* Make sure to add `define( 'BOOST_CACHE', true );` in your `wp-config.php` to get the toggle to show up;
* Make sure that the UI behaves as you'd expect it to. For example, if there's an error during setup, it should show. Fixing it and clicking the "re-run setup" should remove the notice or maybe show a different error.

dilirity committed:
Add toggle UI for Cache

dilirity committed:
add changelog

dilirity committed:
Add health check component for Cache feature

dilirity committed:
Add diagnostic check and UI to show any issues for Page Cache

dilirity committed:
Update canBeEnabled check to check cache signature as well

dilirity committed:
Update canBeEnabled check to actually respect status of module

dilirity committed:
Add comment

dilirity committed:
Add all error messages to diagnostic dictionary

dilirity committed:
Add todo comments

Commit:
Merge branch 'trunk' into add/boost/cache-feature-ui

thingalon commented on projects/plugins/boost/app/assets/src/js/features/module/module.tsx:
```diff
@@ -7,6 +7,7 @@ type ModuleProps = {
 	title: React.ReactNode;
 	description: React.ReactNode;
 	children?: React.ReactNode;
+	alwaysRenderableChildren?: React.ReactNode;
```
This needs a comment describing what it means, and what it's for. It's a confusing name. 

But at the same time, I struggle to come up with a better one. :-) So I recommend a comment.

dilirity replied:
I've removed this since the notice is now going to be a child that will be visible only when toggled.

thingalon commented on projects/plugins/boost/app/assets/src/js/features/page-cache/health/health.tsx:
```diff
@@ -0,0 +1,25 @@
+		}
+	}
+
+	// Don't mind me.
```
LOL

I thought React got upset if you didn't return `null` here. I guess `undefined` is as good as a `null` :)

dilirity replied:
Oh you're right! It either needs to return something or `null`. I've updated the code.

thingalon reviewed (commented):
Looks like the design discussion has moved away from always showing the diagnostic / status information on the Boost Cache module: p1HpG7-r0o-p2#comment-69705

Given that I'll save the review for after this inevitably gets a bit of reworking done. :) 

🤔 Now that the "cacheability" / diagnostics are only going to be used/visible while the module is on, I wonder if it'd make sense to shift from calling it diagnostics, to calling _status_. Where one possible status is "error" with information about why the cache isn't working.

Later we'll likely want to add info like how many pages are in the cache and whatnot. Starting with a status object gives us the scope to add that stuff as we go :)

dilirity committed:
Make sure health component always returns something

dilirity committed:
Merge branch 'trunk' into add/boost/cache-feature-ui

dilirity committed:
Move setup/cleanup logic to a separate place

dilirity committed:
Add way to invalidate queries

dilirity committed:
add changelog

dilirity committed:
Update setup to happen once Cache is enabled;
Remove Diagnostic abstraction, as it's all handled by Setup now;
Update data store to store only the error code
Add fallback error code

dilirity committed:
Add button to re-run Boost Cache setup when error was reported

dilirity committed:
Remove unused property

dilirity committed:
Update variable names

dilirity committed:
Add activate/deactivate traits to modules

dilirity committed:
Add cleanup on page cache deactivation

dilirity committed:
Remove return type

dilirity committed:
Cleanup page cache error store on module deactivation

dilirity commented:
@thingalon I've addressed the things from your review, but I've made a lot of changes based on the design discussions we had.

donnchawp reviewed (approved):
It appears to work ok. I got the stand-in error message when I created a dummy advanced-cache.php
That file and WP_CACHE were added correctly. It changed it from false to true when I tried that.

Looks good. Fixes to the warning message in a later PR?

thingalon commented on projects/plugins/boost/app/assets/src/js/pages/index/index.tsx:
```diff
@@ -31,6 +33,12 @@ const Index = () => {
 
 	const premiumFeatures = usePremiumFeatures();
 
+	// When page cache is enabled, page cache error needs to be invalidated,
+	// so we can get the updated error message from the last setup run.
+	const invalidatePageCacheError = () => {
+		invalidateQuery( 'page_cache_error' );
+	};
```
Only `stores/page-cache.ts` should know what the cache key / datasync key is for the page cache error datastore. Maybe move this to a function in `stores/page-cache.ts`.

dilirity replied:
Oh good catch! I tried a bunch of things and I guess I forgot to move that 😅

thingalon commented on projects/plugins/boost/app/assets/src/js/pages/index/index.tsx:
```diff
@@ -116,6 +124,22 @@ const Index = () => {
+					</p>
+				}
+				onEnable={ invalidatePageCacheError }
+				onDisable={ invalidatePageCacheError }
```
Is it worth invalidating on disable?

dilirity replied:
Yeah. Otherwise, if there was an error when you enabled the module and then disabled it, when you enable again, you'd see the error appear for a sec and disappear, since the error store hasn't been updated yet.

![CleanShot 2024-02-14 at 13 10 29](https://github.com/Automattic/jetpack/assets/11799079/02fd1bec-270b-4aa3-8602-2e04afcdaf60)


I feel like this is a band-aid for a problem I haven't pinpointed yet.

thingalon commented on projects/plugins/boost/app/modules/Modules_Index.php:
```diff
@@ -74,6 +74,18 @@ public function available_modules() {
 		return $available_modules;
 	}
 
+	public function is_module_enabled( $slug ) {
```
I'm surprised no such function existed before :-)

thingalon commented on projects/plugins/boost/app/modules/cache/Page_Cache.php:
```diff
@@ -34,173 +36,24 @@ class Page_Cache implements Pluggable, Is_Always_On {
-	 * These are used by WordPress to load the caching system before most of
-	 * WordPress is loaded.
-	 */
-	public function setup() {
```
Moving this all out to a special setup class responsible for setting up and tearing down was a nice idea 👍

thingalon commented on projects/plugins/boost/app/modules/cache/Page_Cache_Setup.php:
```diff
@@ -0,0 +1,186 @@
+	/**
+	 * Returns true if the wp-content directory is writeable.
+	 */
+	private static function is_wp_content_writable() {
```
Minor nitpick: IMHO `is_` functions should only ever return `true` or `false` and should be expected to be used any time you want to find out something about a system.

Perhaps call it `verify_wp_content_writable` or `test_wp_content_writable`. Not a blocker to this PR.

dilirity replied:
Thanks! I've updated this to `verify_wp_content_writable`. I guess the confusion for me comes from returning `true` - thus I added the prefix `is_`.

thingalon commented on projects/plugins/boost/app/modules/cache/Page_Cache_Setup.php:
```diff
@@ -0,0 +1,186 @@
+			$content = file_get_contents( $advanced_cache_filename ); // phpcs:ignore WordPress.WP.AlternativeFunctions.file_get_contents_file_get_contents
+
+			if ( strpos( $content, Page_Cache::ADVANCED_CACHE_SIGNATURE ) === false ) {
+				return new \WP_Error( 'advanced-cache-incompatible', 'advanced-cache.php exists, but belongs to another plugin/system.' );
```
I like that these WP_Errors each have a code that let our UI describe what's going on in detail.

thingalon commented on projects/plugins/boost/app/modules/cache/data-sync-actions/run-setup.php:
```diff
@@ -0,0 +1,22 @@
+	 * @param mixed            $_data    JSON Data passed to the action.
+	 * @param \WP_REST_Request $_request The request object.
+	 */
+	public function handle( $_data, $_request ) {
+		return Page_Cache_Setup::run_setup();
+	}
```
I like an API endpoint that is short and readable.

thingalon reviewed (approved):
This is not finished yet, as mentioned in the PR comment at the top. It's approved on the basis that the UI can be iterated on from here, and this is a good start.

I have given some inline improvements I'd like to see in the next PR. The big ones imho are: 
- `is_` -> something else, and
- Don't make code outside the store .ts file need to know the cache key to invalidate.

dilirity commented:
Thanks for the reviews guys. I'll be sure to address anything left here in the followup PR.
